src/ast/InheritanceForest.java

```diff
@@ -75,21 +83,64 @@ public InheritanceForest(Program prog) {
 		}
 		
 		LLclasses.node curr;
+		String superName;
+		//meant for semantic check
+		int nodeMapPrevSize = nodeMap.size();
+
 		/*add all other classes to the forest*/
 		while(classBag.head != null) {
-			if(nodeMap.containsKey(classBag.head.value.superName())){
+			superName = classBag.head.value.superName();
+			//semantic check - can't extend main class
+			if(superName.equals(mainClass.name())){
+				isLegalTree = false;
+				return;
+			}
+			if(nodeMap.containsKey(superName)){
+				//semantic check - super is defined before child
+				if(nodeMap.get(superName).value.lineNumber >= classBag.head.value.lineNumber){
+					isLegalTree = false;
+					return;
+				}
+				//semantic check - no name repetition
+				if(nodeMap.containsKey(classBag.head.value.name())){
+					isLegalTree = false;
+					return;
+				}
+
 				addToForest(classBag.head.value);
 				classBag.removeHead();
 				continue;
 			}
 			curr = classBag.head;
 			while(curr.next != null) {
-				if(nodeMap.containsKey(curr.next.value.superName())) {
+				superName = curr.next.value.superName();
+				if(nodeMap.containsKey(superName)) {
+					//semantic check - super is defined before child
+					if(nodeMap.get(superName).value.lineNumber >= curr.next.value.lineNumber){
+						isLegalTree = false;
+						return;
+					}
+					//semantic check - no name repetition
+					if(nodeMap.containsKey(curr.next.value.name())){
+						isLegalTree = false;
+						return;
+					}
+
```

Comment:
this bit is a duplicate of the one before that. I think they can be done with one loop